src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt

```diff
@@ -0,0 +1,566 @@
+                            }
+                        } else {
+                            allItems[name] = pair
+                        }
+                    }
+                }
+            }
+//             print(this)
+        }
+
+        fun getItems(): MutableMap<NeuInternalName, Pair<Double, Int>> {
+            return allItems
+        }
+
+        override fun toString(): String {
+            var result = "ItemsOverall("
+
+            for ((item, pair) in allItems) {
+                result += ("\nItem: $item, Amount: ${pair.first}, in items: ${pair.second}")
+            }
+
+            result += "\n)"
+            return result
+        }
+
+        fun get(item: NeuInternalName) = allItems[item]
+    }
+
+    // TODO soon: somehow also make it searchable?
+    private var display = listOf<Renderable>()
+
+    private var inventoryOpen = false
+
+    var currentlyOpenRecipe: PrimitiveRecipe? = null
+    var displayItem: ItemStack? = null
+
+    // all the functions for interacting with the shopping list come here
+    fun add(itemName: NeuInternalName, amount: Double = 1.0, categoryName: String? = null) {
+        if (!isConfigLoaded) return
+        val items = items
+
+        // TODO: shouldn't happen @Thunderblade73
+        if (!isEnabled()) return
+//         println("Adding ${itemName.itemName} x$amount to $categoryName")
+
+        val category: ShoppingListCategory
+        if (categoryName != null) {
+            if (!categoryName.isCategory()) {
+                category = ShoppingListCategory(categoryName)
+                categories.add(category)
+            } else {
+                category = categories.firstOrNull { it.name == categoryName } ?: return
+            }
+        } else {
+            category = items
+        }
+        category.add(itemName, amount)
+
+        createDisplay()
+    }
+
+    fun addCategory(categoryName: String, color: LorenzColor? = null, saveInStorage: Boolean = true) {
+        if (!isEnabled()) return
+        if (!isConfigLoaded) return
+
+        if (categories.any { it.name == categoryName }) return
+
+        if (color == null) {
+            categories.add(ShoppingListCategory(categoryName, saveInStorage = saveInStorage))
+        } else {
+            categories.add(ShoppingListCategory(categoryName, color = color, saveInStorage = saveInStorage))
+        }
+
+        update()
+    }
+
+    fun removeCategory(categoryName: String) {
+        if (!isEnabled()) return
+        if (!isConfigLoaded) return
+
+        val category = categories.firstOrNull { it.name == categoryName } ?: return
+        categories.remove(category)
+        update()
+    }
+
+    fun removeCategory(category: ShoppingListCategory) {
+        if (!isEnabled()) return
+        if (!isConfigLoaded) return
+
+        categories.remove(category)
+        update()
+    }
+
+    // maybe name it removeCommand ???
+    fun remove(name: String, amount: Double? = null, categoryName: String? = null) {
```

Comment:
This function is 3-4 nested if statements - refactor it into some helper functions if necessary, but this ain't it